src/issues/shareProviders.ts

```diff
@@ -0,0 +1,242 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+import { Commit, Remote, Repository } from '../api/api';
+import { GitApiImpl } from '../api/api1';
+import { fromReviewUri, Schemes } from '../common/uri';
+import { FolderRepositoryManager } from '../github/folderRepositoryManager';
+import { RepositoriesManager } from '../github/repositoriesManager';
+import { getRepositoryForFile } from '../github/utils';
+import { encodeURIComponentExceptSlashes, getBestPossibleUpstream, getOwnerAndRepo, getSimpleUpstream, getUpstreamOrigin, rangeString } from './util';
+
+export class ShareProviderManager implements vscode.Disposable {
+	private disposables: vscode.Disposable[] = [];
+
+	constructor(repositoryManager: RepositoriesManager, gitAPI: GitApiImpl) {
+		if (!vscode.window.registerShareProvider) {
+			return;
+		}
+
+		this.disposables.push(
+			new GitHubDevShareProvider(repositoryManager, gitAPI),
+			new GitHubPermalinkShareProvider(repositoryManager, gitAPI),
+			new GitHubHeadLinkShareProvider(repositoryManager, gitAPI)
+		);
+	}
+
+	dispose() {
+		this.disposables.forEach((d) => d.dispose());
+	}
+}
+
+const supportedSchemes = [Schemes.File, Schemes.Review, Schemes.Pr, Schemes.PRNode, Schemes.FileChange, Schemes.GithubPr, Schemes.VscodeVfs, Schemes.Comment];
+
+abstract class AbstractShareProvider implements vscode.Disposable, vscode.ShareProvider {
+	private disposables: vscode.Disposable[] = [];
+	protected shareProviderRegistrations: vscode.Disposable[] | undefined;
+
+	constructor(
+		protected repositoryManager: RepositoriesManager,
+		protected gitAPI: GitApiImpl,
+		public readonly id: string,
+		public readonly label: string,
+		public readonly priority: number,
+		private readonly origin = 'github.com'
+	) {
+		this.initialize();
+	}
+
+	public dispose() {
+		this.disposables.forEach((d) => d.dispose());
+		this.shareProviderRegistrations?.map((d) => d.dispose());
+	}
+
+	private async initialize() {
+		if ((await this.hasGitHubRepositories()) && this.shouldRegister()) {
+			this.register();
+		} else {
+			this.disposables.push(this.gitAPI.onDidOpenRepository(async e => {
+				await e.status();
+				if ((await this.hasGitHubRepositories()) && this.shouldRegister()) {
+					this.register();
+				}
+			}));
```

Comment:
You can use the folder repo manager events for this: 

https://github.com/microsoft/vscode-pull-request-github/blob/7ea3684f3b991402df8995a505eaca828101976d/src/github/folderRepositoryManager.ts#L150-L154

I think you should be able to get rid of the `status()` call here and in `hasGitHubRepositories` if you use the folder repo manager events.

Reply:
`onDidLoad` is exactly what I needed! I think I need a specific URI in order to know which folder repo manager to use, so I'm using the equivalent `onDidLoadAnyRepository` event on the parent `RepositoriesManager` class.